README.md

```diff
@@ -23,17 +23,17 @@ const {
   AOHost, PingPong, Iceberg, TWAP, AccumulateDistribute, MACrossover
 } = require('bfx-hf-algo')
 
-const HFDB = require('bfx-hf-models')	
-const HFDBLowDBAdapter = require('bfx-hf-models-adapter-lowdb')	
-const {	
```

Comment:
interesting!

for a 40 seconds timeframe you can see how memory usage is slowly increasing, even with the current api limitations

<img width="725" alt="Screenshot 2021-05-18 at 10 48 05" src="https://user-images.githubusercontent.com/298166/118621502-b7345580-b7c6-11eb-9347-67ca353257de.png">

Reply:
<img width="692" alt="Screenshot 2021-05-18 at 10 52 15" src="https://user-images.githubusercontent.com/298166/118621954-2611ae80-b7c7-11eb-9c8c-440b98d75e99.png">

Reply:
Since it returns error 429 in the current implementation, for some reason Garbage collector doesn't clear these error objects. The following code produces an unhandledPromiseRejection

```js
setInterval(async () => {
  const host = spawnHost()

  await connectHost(host)
  await destroyHost(host)
}, 100)
```

![image](https://user-images.githubusercontent.com/70510980/118707693-81718a00-b823-11eb-9099-b1956709aee7.png)

Even if we ignore them, some data still remain in memory

```js
setInterval(async () => {
  const host = spawnHost()

  try {
    await connectHost(host)
    await destroyHost(host)
  } catch (e) {}
}, 100)
```

![image](https://user-images.githubusercontent.com/70510980/118708354-4de32f80-b824-11eb-89f8-8f273320ff54.png)

Reply:
After deleting all listeners (`this.m.on` in `bfx-hf-algo/lib/ws_adapter.js`, `adapter.on` in `bfx-hf-algo/lib/host/ws2/bind_bus.js`, etc)

![image](https://user-images.githubusercontent.com/70510980/118868820-53587c80-b8ed-11eb-962d-71620f497ac6.png)

Reply:
For examples, `(closure)` section contains only the external module `ws` among all entries 

![image](https://user-images.githubusercontent.com/70510980/118870827-9f0c2580-b8ef-11eb-8bda-1e8060234984.png)

Reply:
![image](https://user-images.githubusercontent.com/70510980/118965108-f900ff80-b970-11eb-940e-597f66547f56.png)

Reply:
At the current stage of research I came to this code which is used in `bfx-api-node-core` dependency and it leads to the same memory leak


```js
const initWSState = require('bfx-api-node-core/lib/ws2/init_state')
const { WS_URL } = require('bfx-api-node-core/lib/config')
const WebSocket = require('ws')

/// setInterval here
/////

  const {
    apiKey,
    apiSecret
  } = wsSettings

  let wsState = null

  try {
    wsState = initWSState({
      url: WS_URL,
      apiKey,
      apiSecret,
      plugins: []
    })

  } catch (e) {
    console.log(e)
  } finally {
    if (!wsState) return
    const { ws } = wsState

    if (ws.readyState !== WebSocket.CLOSED) {
      ws.close()
    }
  }
```

![image](https://user-images.githubusercontent.com/70510980/119029205-050cb180-b9b1-11eb-93c6-424e9d2a2d74.png)

Reply:
Simplifed example:

```js
setInterval(async () => {
  const ws = new WebSocket('wss://api.bitfinex.com/ws/')
  ws.on('error', () => {})
  if (ws.readyState !== WebSocket.CLOSED) {
    ws.close()
  }
  ws.terminate()
}, 30)
```

Reply:
Created the issue https://github.com/websockets/ws/issues/1889

Reply:
while preparing food it came to my mind:

```js
ws.on('error', () => {})
```

a `removeListener` is missing...

Reply:
If the event isn't bound to the global scope, the handler will be removed along with `ws` object